commons/zenoh-shm/src/api/protocol_implementations/posix/posix_shared_memory_provider_backend.rs

```diff
@@ -0,0 +1,284 @@
+                            size: chunk.size - required_len,
+                        };
+                        tracing::trace!(
+                            "The allocation will leave a Free Chunk: {:?}",
+                            &free_chunk
+                        );
+                        guard.push(free_chunk);
+                        chunk.size = required_len;
+                    }
+                    self.available.fetch_sub(chunk.size, Ordering::Relaxed);
+
+                    let descriptor =
+                        ChunkDescriptor::new(self.segment.segment.id(), chunk.offset, chunk.size);
+
+                    Ok(AllocatedChunk {
+                        descriptor,
+                        data: unsafe {
+                            AtomicPtr::new(self.segment.segment.elem_mut(chunk.offset))
+                        },
+                    })
+                }
+                Some(c) => {
+                    tracing::trace!("PosixSharedMemoryProviderBackend::alloc({:?}) cannot find any big enough chunk\nSharedMemoryManager::free_list = {:?}", layout, self.free_list);
+                    guard.push(c);
+                    Err(ZAllocError::NeedDefragment)
+                }
+                None => {
+                    // NOTE: that should never happen! If this happens - there is a critical bug somewhere around!
+                    let err = format!("PosixSharedMemoryProviderBackend::alloc({:?}) cannot find any available chunk\nSharedMemoryManager::free_list = {:?}", layout, self.free_list);
+                    #[cfg(feature = "test")]
+                    panic!("{err}");
+                    #[cfg(not(feature = "test"))]
+                    {
+                        tracing::error!("{err}");
+                        Err(ZAllocError::OutOfMemory)
+                    }
+                }
+            }
+        } else {
+            tracing::trace!( "PosixSharedMemoryProviderBackend does not have sufficient free memory to allocate {:?}, try de-fragmenting!", layout);
+            Err(ZAllocError::OutOfMemory)
+        }
+    }
+
+    fn free(&self, chunk: &ChunkDescriptor) {
+        let free_chunk = Chunk {
+            offset: chunk.chunk,
+            size: chunk.len,
+        };
+        self.available.fetch_add(free_chunk.size, Ordering::Relaxed);
+        self.free_list.lock().unwrap().push(free_chunk);
+    }
+
+    fn defragment(&self) -> usize {
+        fn try_merge_adjacent_chunks(a: &Chunk, b: &Chunk) -> Option<Chunk> {
+            let end_offset = a.offset as usize + a.size;
+            if end_offset == b.offset as usize {
+                Some(Chunk {
+                    size: a.size + b.size,
+                    offset: a.offset,
+                })
+            } else {
+                None
+            }
+        }
+
+        let mut largest = 0usize;
+
+        let mut guard = self.free_list.lock().unwrap();
+        if guard.len() > 1 {
```

Comment:
In general is good to add a comment explaining the algorithm. It makes easier the review as well.